### PR TITLE
CDAP-14617 Write tms messages for different topics to different LevelDB tables

### PR DESCRIPTION
**Pull request by anwar6953**
https://issues.cask.co/browse/CDAP-14617
https://builds.cask.co/browse/CDAP-DUT6676-3

### CODE REVIEW

File: cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java [hunk 1]
```diff
@@ -69,42 +81,76 @@ public LevelDBTableFactory(CConfiguration cConf) {
     executor.scheduleAtFixedRate(new DataCleanup(), 0L,
                                  Long.parseLong(cConf.get(Constants.MessagingSystem.LOCAL_DATA_CLEANUP_FREQUENCY)),
                                  TimeUnit.SECONDS);
+
+    // Currently we don't support customizable table name yet, hence always get it from cConf.
+    // Later on it can be done by topic properties, with impersonation setting as well.
```
Comment (chtyim) [inline cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java]: We won't have impersonation in LevelDB, right?

File: cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java [hunk 2]
```diff
@@ -69,42 +81,76 @@ public LevelDBTableFactory(CConfiguration cConf) {
...
+    this.metadataTableName = cConf.get(Constants.MessagingSystem.METADATA_TABLE_NAME);
+    this.messageTableName = cConf.get(Constants.MessagingSystem.MESSAGE_TABLE_NAME);
+    this.payloadTableName = cConf.get(Constants.MessagingSystem.PAYLOAD_TABLE_NAME);
+    this.messageTableCache = CacheBuilder.newBuilder().build(new CacheLoader<TopicMetadata, LevelDBMessageTable>() {
```
Comment (chtyim) [inline cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java]: The cache should have an expiration and a removal listener

File: cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java [hunk 3]
```diff
@@ -69,42 +81,76 @@ public LevelDBTableFactory(CConfiguration cConf) {
...
+      @ParametersAreNonnullByDefault
+      @Override
+      public LevelDBMessageTable load(TopicMetadata key) throws Exception {
+        File dbPath = getDataDBPath(messageTableName, key.getTopicId(), key.getGeneration());
+        LevelDBMessageTable messageTable =
+          new LevelDBMessageTable(LEVEL_DB_FACTORY.open(dbPath, dbOptions), key);
+        LOG.info("Messaging message table created at {}", dbPath);
```
Comment (chtyim) [inline cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java]: Log as debug

File: cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java [hunk 4]
```diff
@@ -69,42 +81,76 @@ public LevelDBTableFactory(CConfiguration cConf) {
...
-  public synchronized MessageTable createMessageTable(String tableName) throws IOException {
-    if (messageTable != null) {
-      return messageTable;
-    }
-
-    File dbPath = ensureDirExists(new File(baseDir, NamespaceId.SYSTEM.getNamespace() + "." + tableName));
-    messageTable = new LevelDBMessageTable(LEVEL_DB_FACTORY.open(dbPath, dbOptions));
-    LOG.info("Messaging message table created at {}", dbPath);
-    return messageTable;
+  public synchronized MessageTable createMessageTable(TopicMetadata topicMetadata) {
```
Comment (chtyim) [inline cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java]: I don't think you need `synchronized` anymore

File: cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java [hunk 5]
```diff
@@ -118,16 +164,48 @@ private File ensureDirExists(File dir) throws IOException {
...
+          List<File> filesToDelete = new ArrayList<>();
+          for (int olderGeneration = cleanOlderThan - 1; olderGeneration > 0; olderGeneration--) {
+            File dataDBPath = getDataDBPath(messageTableName, metadata.getTopicId(), olderGeneration, false);
+            if (!dataDBPath.exists()) {
+              break;
+            }
+            filesToDelete.add(dataDBPath);
+          }
+
+          for (int i = filesToDelete.size() - 1; i >= 0; i--) {
```
Comment (chtyim) [inline cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java]: Use a `Deque` instead, then you can use normal iterator.

File: cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java [hunk 6]
```diff
@@ -118,16 +164,48 @@ private File ensureDirExists(File dir) throws IOException {
...
+        }
+      } catch (IOException ex) {
+        LOG.debug("Unable to perform data cleanup in TMS LevelDB tables", ex);
+      }
+
+      long timeStamp = System.currentTimeMillis();
+      try {
+        for (Map.Entry<TopicMetadata, LevelDBMessageTable> messageTableEntry : messageTableCache.asMap().entrySet()) {
+          messageTableEntry.getValue().pruneMessages(messageTableEntry.getKey(), timeStamp);
+          System.out.println(messageTableEntry.getKey());
```
Comment (chtyim) [inline cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java]: No `System.out`

File: cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java (hunk 1 above)
Comment (anwar6953) [reply cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java]: right, moved it to the proper spot.

File: cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java (hunk 6 above)
Comment (anwar6953) [reply cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBTableFactory.java]: oops, forgot to remove this.

**Review by chtyim — APPROVED** [review]
LGTM

**Comment by anwar6953** [comment]
Filed https://issues.cask.co/browse/CDAP-14725 as a followup.

**Pull request merged**